src/main/java/org/prebid/server/bidder/pubmatic/PubmaticBidder.java

```diff
@@ -1,30 +1,252 @@
+        }
+        return extImpPubmatic.getWrapExt();
+    }
+
+    private static Imp parseAndValidateImp(Imp imp, ExtImpPubmatic extImpPubmatic) {
+        if (imp.getBanner() == null && imp.getVideo() == null) {
+            throw new PreBidException(String.format("Invalid MediaType. PubMatic only supports Banner and Video. "
+                    + "Ignoring ImpID=%s", imp.getId()));
+        }
+
+        final Imp.ImpBuilder result = imp.toBuilder();
+        if (imp.getAudio() != null) {
+            result.audio(null);
+        }
+
+        if (imp.getBanner() != null) {
+            final String adSlotString = extImpPubmatic.getAdSlot().trim();
+            final String[] adSlot = adSlotString.split("@");
+            if (adSlot.length != 2 || StringUtils.isBlank(adSlot[0]) || StringUtils.isBlank(adSlot[1])) {
+                throw new PreBidException("Invalid adSlot provided");
+            }
+            result.tagid(adSlot[0]);
+
+            final String[] adSize = adSlot[1].trim().toLowerCase().split("x");
+            if (adSize.length != 2) {
+                throw new PreBidException("Invalid size provided in adSlot");
+            }
+
+            Integer width;
+            Integer height;
+            final String[] heightStr = adSize[1].trim().split(":");
+            try {
+                width = Integer.valueOf(adSize[0].trim());
+                height = Integer.valueOf(heightStr[0].trim());
+            } catch (NumberFormatException e) {
+                throw new PreBidException("Invalid width or height provided in adSlot");
+            }
+            final Banner updatedBanner = imp.getBanner().toBuilder().w(width).h(height).build();
+            result.banner(updatedBanner);
+        }
+
+        if (CollectionUtils.isNotEmpty(extImpPubmatic.getKeywords())) {
+            result.ext(makeKeywords(extImpPubmatic.getKeywords()));
+        } else {
+            result.ext(null);
+        }
+        return result.build();
+    }
+
+    private static ObjectNode makeKeywords(List<ExtImpPubmaticKeyVal> keywords) {
+        List<String> eachKv = new ArrayList<>();
+        for (ExtImpPubmaticKeyVal keyVal : keywords) {
+            if (CollectionUtils.isEmpty(keyVal.getValue())) {
+                logger.error(String.format("No values present for key = %s", keyVal.getKey()));
+            } else {
+                eachKv.add(String.format("\"%s\":\"%s\"", keyVal.getKey(),
+                        String.join(",", keyVal.getValue())));
+            }
+        }
+        final String keywordsString = "{" + String.join(",", eachKv) + "}";
+        try {
+            return Json.mapper.readValue(keywordsString, ObjectNode.class);
+        } catch (IOException e) {
+            throw new PreBidException(String.format("Failed to create keywords with error: %s", e.getMessage()), e);
+        }
+    }
+
+    private static ExtImpPubmatic parsePubmaticExt(Imp imp) {
+        ExtImpPubmatic extImpPubmatic;
```

Comment:
1. Can be refactored to:
```
    private static ExtImpPubmatic parsePubmaticExt(Imp imp) {
        try {
            return Json.mapper.<ExtPrebid<?, ExtImpPubmatic>>convertValue(imp.getExt(),
                    PUBMATIC_EXT_TYPE_REFERENCE).getBidder();
        } catch (IllegalArgumentException e) {
            throw new PreBidException(String.format("Failed to parse imp.ext with error: %s", e.getMessage()), e);
        }
    }
```
Since we have single action and throwing error.

2. Make sure methods ordering: put this method after `makeHttpRequests(...)`. Check all others ordering.